WordPress/Classes/ViewRelated/NUX/LoginEmailViewController.swift

```diff
@@ -1,35 +1,328 @@
+        adjustViewForKeyboard(SigninEditingState.signinEditingStateActive)
+        if SigninEditingState.signinEditingStateActive {
+            emailTextField.becomeFirstResponder()
+        }
+    }
+
+
+    // MARK: - Instance Methods
+
+
+    /// Makes the call to retrieve Safari shared credentials if they exist.
+    ///
+    func fetchSharedWebCredentialsIfAvailable() {
+        didRequestSafariSharedCredentials = true
+        SigninHelpers.requestSharedWebCredentials { [weak self] (found, username, password) in
+            self?.handleFetchedWebCredentials(found, username: username, password: password)
+        }
+    }
+
+
+    /// Handles Safari shared credentials if any where found.
+    ///
+    /// - Parameters:
+    ///     - found: True if credentails were found.
+    ///     - username: The selected username or nil.
+    ///     - password: The selected password or nil.
+    ///
+    func handleFetchedWebCredentials(_ found: Bool, username: String?, password: String?) {
+        didFindSafariSharedCredentials = found
+
+        guard let username = username, let password = password else {
+            return
+        }
+
+        // Update the login fields
+        loginFields.username = username
+        loginFields.password = password
+
+        // Persist credentials as autofilled credentials so we can update them later if needed.
+        loginFields.safariStoredUsernameHash = username.hash
+        loginFields.safariStoredPasswordHash = password.hash
+
+        signinWithUsernamePassword(false)
+
+        WPAppAnalytics.track(.loginAutoFillCredentialsFilled)
+    }
+
+
+    /// Displays the wpcom sign in form, optionally telling it to immedately make
+    /// the call to authenticate with the available credentials.
+    ///
+    /// - Parameters:
+    ///     - immediateSignin: True if the newly loaded controller should immedately attempt
+    ///                        to authenticate the user with the available credentails.  Default is `false`.
+    ///
+    func signinWithUsernamePassword(_ immediateSignin: Bool = false) {
         performSegue(withIdentifier: .showWPComLogin, sender: self)
     }
 
-    override func signinToSelfHostedSite() {
+
+    /// Displays the self-hosted sign in form.
+    ///
+    func signinToSelfHostedSite() {
         performSegue(withIdentifier: .showSelfHostedLogin, sender: self)
     }
+
+
+    /// Proceeds along the "magic link" sign-in flow, showing a form that let's
+    /// the user request a magic link.
+    ///
+    func requestLink() {
+        performSegue(withIdentifier: .startMagicLinkFlow, sender: self)
+    }
+
+
+    /// Validates what is entered in the various form fields and, if valid,
+    /// proceeds with the submit action.
+    ///
+    func validateForm() {
+        displayError(message: "")
+        guard loginFields.username.isValidEmail() else {
+            assertionFailure("Form should not be submitted unless there is a valid looking email entered.")
```

Comment:
The form is still submittable by using a keyboard by hitting Enter.

When that happens and an email hasn't been entered, the assertion fails.